### ISSUE TITLE
Applying dirichlet BC conditions in example 18 

### ISSUE THREAD

**Issue by makrandak** [mfem/mfem#1774]
Hello all,

 I am trying to apply dirichlet BC condiitons to the DG discretization of Euler equations using the basic template of ex18.  

I modified example18 using the following process. 
a) I changed the mesh from periodic to "inline-quad.mesh" 
b) I populated the array attributes vector `array<int> attr(pmesh.bdr_attributes.Max());` appropriately for each deg of freedom (4 or 5 depending on the no of spatial dimensions.)
c) I also populate a vector for each deg of freedom which contains `attr` and `Coefficient* coeff` (a function to evaluate the value on the boundary depending on the coordinate.)
d) 	Then I Get the essential dof from the finite element space containing all deg of freedom ( `vfes`in the example) 
```c++
finiteElementSpace->GetEssentialTrueDofs(density_ess_attr_, density_ess_tdof_, 0);
finiteElementSpace->GetEssentialTrueDofs(momentumX_ess_attr_, momentumX_ess_tdof_, 1);
finiteElementSpace->GetEssentialTrueDofs(momentumY_ess_attr_, momentumY_ess_tdof_, 2);
finiteElementSpace->GetEssentialTrueDofs(energy_ess_attr_, energy_ess_tdof_, 3);
```

I am right now a bit confused as to where I should really provide these data structures in the finite element space in example18. 
Any help on where I should supply these data structures would be much appreciated.

**Comment by pazner** [comment]
You could take a look at issue #1636, where boundary conditions for the Sod shock tube problem were discussed. Generally speaking, the boundary conditions in this example should be enforced weakly by modifying the boundary integrators, and not by enforcing "essential" boundary conditions.

**Comment by makrandak** [comment]
Thank you very much Will for your response.  I am a new user of mfem and still a bit confused about the `bdr_attr` strategy.  I have two follow up questions.  

1) I understand from reading issue #1636 that weak imposition of the boundary conditions can be done using face integrators.  The boundary markers (`Array<int> &bdr_marker`) specified in the `AddBdFaceIntegrator` method, are they specific to each deg of freedom ? I anticipate having different dirichlet BC on different boundary attributes for each deg of freedom.

I am asking this because I am trying to create a generic way to apply BC for different cases, sod shock would be one of them.  

2) Strong imposition of boundary conditions should also be an option right, if that is the route I want to take, how should I do it? Even if this may induce some oscillations near the boundary.  I feel that this should also be an option, then I can compare weak and strong imposition of BC. 

If there is an example for a codebase I can look at which does Dirichlet BC for Euler that would be great. Even if there is an example where DG method is used with multiple deg of freedom and Dirichlet BC are used, that would be great.

**Comment by pazner** [comment]
> 1. I understand from reading issue #1636 that weak imposition of the boundary conditions can be done using face integrators.  The boundary markers (`Array<int> &bdr_marker`) specified in the `AddBdFaceIntegrator` method, are they specific to each deg of freedom ? I anticipate having different dirichlet BC on different boundary attributes for each deg of freedom.

The `bdr_marker` array is used to specify the different parts of the boundary where you want to enforce different conditions. These are not per degree of freedom, but rather associated with different geometric subsets of the mesh boundary.

I'm not sure what you mean by having different boundary conditions for each degree of freedom. It would be helpful if you included more specific information about your problem.

> 2. Strong imposition of boundary conditions should also be an option right, if that is the route I want to take, how should I do it? Even if this may induce some oscillations near the boundary.  I feel that this should also be an option, then I can compare weak and strong imposition of BC.

You can of course use the list of DOFs returned by `GetEssentialTrueDofs` to set the grid function to strongly satisfy your given BCs, but I do not recommend doing this with DG discretizations.

> If there is an example for a codebase I can look at which does Dirichlet BC for Euler that would be great. Even if there is an example where DG method is used with multiple deg of freedom and Dirichlet BC are used, that would be great.

I agree this would be useful (and such an example probably exists), but I don't have a reference off-hand.

**Comment by makrandak** [comment]
Hello Will,  Thanks for your response. 
> 
> The `bdr_marker` array is used to specify the different parts of the boundary where you want to enforce different conditions. These are not per degree of freedom, but rather associated with different geometric subsets of the mesh boundary.

 I understand now that the whole flux vector is adjusted for a particular boundary condition when being weakly imposed, so I understand now what `bdr_marker` should contain, it contains the markers where the boundary integral needs to be computed. 
> I'm not sure what you mean by having different boundary conditions for each degree of freedom. It would be helpful if you included more specific information about your problem.
> 
To help with the questions, I am trying to reproduce the implementation mentioned in issue #1636.  So the SOD shock problem with `inline-quad` mesh.  I am a bit confused on the implementation of the boundary integrator.   If I understand correctly I have to compute the flux corrected for the dirichlet boundary condition on the boundary.  I am confused because `AssembleFaceVector` which is the method I am overriding takes two elements as input, I am not sure which two elements are these if we are on the boundary? Maybe I am missing something here. If I understand correctly one needs to implement (F.n, w) on the boundary where F is the corrected for the Dirichlet BC using the Riemann solver, this requires only one boundary element.  I am not able to implement this as `AssembleFaceVector` takes two elements.  Also the decision to compute Ur or Ul probably depends on the normal right? 

On the other had if I try to implement (F.n,[w]) like the interior face terms just by modifying interior face integrator, with corrected flux adjusted for Dirichlet BC, when I did this I run into asserts in AddElementVector with the following error
```
Assertion failed: (dofs.Size() == elemvect.Size()) is false:
 --> Size mismatch: length of dofs is 16, length of elemvect is 32
 ... in function: void mfem::Vector::AddElementVector(const mfem::Array<int>&, const mfem::Vector&)
 ... in file: /home/makrand/Packages/blastfemdg/external/mfem/linalg/vector.cpp:575
```

> 
> You can of course use the list of DOFs returned by `GetEssentialTrueDofs` to set the grid function to strongly satisfy your given BCs, but I do not recommend doing this with DG discretizations.

It would be great if I could still implement this, as I need to compare the two options.  Could you tell me a method I could use the list of DOFs returned by component wise `GetEssentialTrueDofs`.  

 
> I agree this would be useful (and such an example probably exists), but I don't have a reference off-hand.
It would be great If you could let me know about this or maybe refer to someone who know about such an example/codebase.
I really think an example would be very beneficial on applying Dirichlet BC in example 18 as there are many people asking these questions, but that is just a suggestion. 

I appreciate your time and help.

**Comment by pazner** [comment]
> To help with the questions, I am trying to reproduce the implementation mentioned in issue #1636. So the SOD shock problem with `inline-quad` mesh. I am a bit confused on the implementation of the boundary integrator. If I understand correctly I have to compute the flux corrected for the dirichlet boundary condition on the boundary. I am confused because `AssembleFaceVector` which is the method I am overriding takes two elements as input, I am not sure which two elements are these if we are on the boundary? Maybe I am missing something here. If I understand correctly one needs to implement (F.n, w) on the boundary where F is the corrected for the Dirichlet BC using the Riemann solver, this requires only one boundary element. I am not able to implement this as `AssembleFaceVector` takes two elements. Also the decision to compute Ur or Ul probably depends on the normal right?

You are correct, `AssembleFaceVector` takes in two elements. In the case of an interior face, these correspond to the two neighboring elements containing the given face. In the case of a boundary face, there is only one element containing the given face, and so the second element is just a dummy element pointing to the first one.

Typically for the Euler equations, `AssembleFaceVector` should compute an (approximate) solution to the Riemann problem involving the states on either side of the face. This can be done using e.g. the Lax-Friedrichs method (this is probably the simplest/easiest, and is implemented in ex18 in the `RiemannSolver` class, but you can also use Roe, or an exact Riemann solver, etc.

A typical way of handing the boundary conditions is to provide a fictitious external state to the Riemann solver. In other words, if the interior state is denoted `uL`, then you can specify an external state `uR`, and return the result `Fhat(uL, uR, n)`, where `Fhat` corresponds to the Riemann solver. It is probably sufficient in your case to set `uR` to be the farfield state.

> On the other had if I try to implement (F.n,[w]) like the interior face terms just by modifying interior face integrator, with corrected flux adjusted for Dirichlet BC, when I did this I run into asserts in AddElementVector with the following error
> 
> ```
> Assertion failed: (dofs.Size() == elemvect.Size()) is false:
>  --> Size mismatch: length of dofs is 16, length of elemvect is 32
>  ... in function: void mfem::Vector::AddElementVector(const mfem::Array<int>&, const mfem::Vector&)
>  ... in file: /home/makrand/Packages/blastfemdg/external/mfem/linalg/vector.cpp:575
> ```

Are you sure this is on the boundary? On the boundary, `elemvect` should contain the residual corresponding to one element, and on an interior face, `elemvect` should contain the residual corresponding to the two neighboring elements. Since in this case your vector is twice as long as the number of DOFs, it looks like this is an interior face.

> It would be great if I could still implement this, as I need to compare the two options. Maybe this is really straightforward but because I am new to mfem I am not sure which function needs to be used and on where to apply the BC after `GetEssentialTrueDofs` is computed.
> Could you tell me a function I could use the list of DOFs returned by component wise `GetEssentialTrueDofs`?

Actually, I realize that this is a bit harder to implement in MFEM than my previous message indicated. Functions like `GetEssentialTrueDofs` and `GridFunction::ProjectBdrCoefficient` will not work properly with DG spaces. In particular, with DG you can choose to use the Gauss-Legendre basis, which does not have nodes lying on the boundary, so the projection becomes nontrivial. If you use Gauss-Lobatto basis, you can set the boundary DOFs, but this might require some manual effort.

**Comment by makrandak** [comment]
I fixed the assert problem the size of elvect was incorrectly set.  I am adding code here so it may help other users looking at the discussion too.  Replicating and modifying the interior face integrator in the example. I implemented the Boundary integrator as follows:
```
void BoundaryFaceIntegrator::AssembleFaceVector(const FiniteElement &el1,
                                                const FiniteElement &el2,
                                                FaceElementTransformations &Tr,
                                                const Vector &elfun, Vector &elvect) {
	// Compute the term <F.n(u),[w]> on the interior faces.
	const int dof1 = el1.GetDof();
	const int dof2 = el2.GetDof();

	shape1.SetSize(dof1);
	shape2.SetSize(dof2);

	elvect.SetSize((dof1) * inputData_->num_equation);
	elvect = 0.0;

	DenseMatrix elfun1_mat(elfun.GetData(), dof1, inputData_->num_equation);
	DenseMatrix elfun2_mat(elfun.GetData() + dof1 * inputData_->num_equation, dof2,
	                       inputData_->num_equation);

	DenseMatrix elvect1_mat(elvect.GetData(), dof1, inputData_->num_equation);
	DenseMatrix elvect2_mat(elvect.GetData() + dof1 * inputData_->num_equation, dof2,
	                        inputData_->num_equation);

	// Integration order calculation from DGTraceIntegrator
	int intorder;
	if (Tr.Elem2No >= 0)
		intorder = (std::min(Tr.Elem1->OrderW(), Tr.Elem2->OrderW()) + 2 * std::max(el1.GetOrder(), el2.GetOrder()));
	else {
		intorder = Tr.Elem1->OrderW() + 2 * el1.GetOrder();
	}
	if (el1.Space() == FunctionSpace::Pk) {
		intorder++;
	}
	const IntegrationRule *ir = &IntRules.Get(Tr.GetGeometryType(), intorder);

	/// loop over integration points
	for (int i = 0; i < ir->GetNPoints(); i++) {
		const IntegrationPoint &ip = ir->IntPoint(i);

		Tr.SetAllIntPoints(&ip); // set face and element int. points

		// Calculate basis functions on both elements at the face
		el1.CalcShape(Tr.GetElement1IntPoint(), shape1);
		el2.CalcShape(Tr.GetElement2IntPoint(), shape2);

		// Interpolate elfun at the point
		elfun1_mat.MultTranspose(shape1, funval1);
		elfun2_mat.MultTranspose(shape2, funval2);

		// Get the normal vector and the flux on the face
		CalcOrtho(Tr.Jacobian(), nor);

                // Calculates funval2 as 2D - funval1, where D is the dirichlet state. 
		this->calcDirichletState(funval1, funval2);
		
                /// Calculate the Fhat using Reimann solver
                const double mcs = rsolver.Eval(funval1, funval2, nor, fluxN);

		// Update max char speed
		if (mcs > inputData_->max_char_speed) { inputData_->max_char_speed = mcs; }

		fluxN *= ip.weight;
		for (int k = 0; k < inputData_->num_equation; k++) {
			for (int s = 0; s < dof1; s++) {
				elvect1_mat(s, k) -= fluxN(k) * shape1(s);
			}
			for (int s = 0; s < dof2; s++) {
				elvect2_mat(s, k) += fluxN(k) * shape2(s);
			}
		}
	}
}
```
This is still calculating (F.n, [w]) at the boundary.  Do you think this is correct.  I am not sure whether I should modify this to (F.n,w) on the boundary. 

I call this in main like this, assuming only application of BC on the left boundary 
```
Array<int> ess_bdr(pmesh.bdr_attributes.Max());
	ess_bdr = 0;
	ess_bdr[3] = 1;
	//ess_bdr[2] = 1;
	std::cout<< "ess_bdr is:  " << ess_bdr[0] << ", " << ess_bdr[1] << ", " << ess_bdr[2] << ", " << ess_bdr[3] <<
	std::endl;
	A.AddBdrFaceIntegrator( new BoundaryFaceIntegrator(rsolver, dim, &inputData, &bc), ess_bdr);
```
Is this correct ?? Here `&inputData` and `&bc` are references to custom objects which carry some information as I am trying to write a general framework to handle multiple cases.  

I am still running into nans when I use this. 

> Actually, I realize that this is a bit harder to implement in MFEM than my previous message indicated. Functions like GetEssentialTrueDofs and GridFunction::ProjectBdrCoefficient will not work properly with DG spaces. In particular, with DG you can choose to use the Gauss-Legendre basis, which does not have nodes lying on the boundary, so the projection becomes nontrivial. If you use Gauss-Lobatto basis, you can set the boundary DOFs, but this might require some manual effort.
>
How do I change to `Gauss Lobatto` basis probably in the constructor on finiteElementspace? I did not know about this function `GridFunction::ProjectBdrCoefficient`, Is there an example I can look at for how to apply these BC for multiple dof and uses `GridFunction::ProjectBdrCoefficient` to apply specific values on the boundaries? 

I am pursuing both options because I want to give users both options weak and strong imposition and I want to do stability analysis and convergence analysis with both weak and strong imposition. 

Thank you very much for your help. I really appreciate it.

**Comment by makrandak** [comment]
Hello Will @pazner,  I was able to implement a general class for handling various boundary conditions weakly using this guidance. Thanks for your help on this.  

I would like to bring one issue to your notice which might be useful to other users.  The `Array<int> boundaryMarkers` is not owned by `NonlinearForm` class so this array has to stay in scope all the time as the boundary conditions may get applied multiple times (every timestep if the bc is time dependent).   

When I was writing a class to abstract out the setup of weak boundary conditions I had to declare `Array<Array<int>> boundaryMarkers;` object in the main which I populated inside the method of this class as follows,
```
BoundaryConditions bc(&pmesh, &inputData, &A, &rsolver, verbose);
Array<Array<int>> boundaryMarkers;
bc.setupBoundaryConditions(boundaryMarkers);
```
as opposed to having a data member inside `BoundaryCondtions` as it would go out of scope.  It is cleaner if the boundaryMarkers is a member of `BoundaryConditions` class because then the `main` of the code is cleaner.    

I am not sure why `boundaryMarkers` is not owned by `NonlinearForm` class.  Could one of you let me know the reason for this?   

I just wanted to let the users know about this, as they might make the same mistake as me, and let it go out of scope producing undefined behaviour. 

I am still not sure how to add the boundary conditions strongly but maybe that is a discussion for a different post. 

Thanks Will @pazner again for your help on this.
